crates/adapters/src/transport/durable/mod.rs

```diff
@@ -0,0 +1,192 @@
+/// it is read again, it will be the same as the first time.
+pub trait DurableInputEndpoint {
+    /// The range of steps that are available to be read.
+    fn steps(&self) -> AnyResult<Range<Step>>;
+
+    /// Opens the endpoint for reading steps and possibly adding them at the
+    /// end.  The reader passes data and status information to `consumer`.
+    fn open(&self, consumer: Box<dyn DurableInputConsumer>) -> Box<dyn DurableInputReader>;
+
+    /// Notifies the endpoint that steps less than `step` aren't needed
+    /// anymore.  It may optionally discard them.
+    fn expire(&self, _step: Step) {}
+}
+
+/// Reader and appender for [`DurableInputTransport`].
+///
+/// A `DurableInputReader` reads from a durable input transport step-by-step.
+/// It first reads the step specified in the first call to [`next`](Self::next),
+/// and advances as far as the highest step passed in any later call to
+/// [`next`](Self::next).  A reader can also cause new steps to be appended, by
+/// calling [`next`](Self::next) with a step number that does not yet exist.
+///
+/// Call [`DurableInputEndpoint::open`] to create a `DurableInputReader`.  The
+/// reader passes the data that it reads to the [consumer](DurableInputConsumer)
+/// passed to the `open` call.
+///
+/// The reader reads and appends asynchronously in separate threads.  Drop the
+/// reader to stop these threads and stop passing data to the consumer.
+///
+/// > 🚨️**Warning**: Only a single appender should be active at once for a given
+/// endpoint.
+pub trait DurableInputReader {
+    /// Requests that the endpoint reads steps up to and including `step`.  If
+    /// this is the first call for a given reader, then `step` will be the first
+    /// step read; otherwise, it requests continuing to read all of the steps up
+    /// to and including `step`.
+    ///
+    /// `step` has the following interpretation given that the input already
+    /// contains steps `first..last`:
+    ///
+    /// - If `step` is in `first..last`, then the reader will read up to and
+    /// including `step` and pass each step's data to the consumer.
+    ///
+    /// - If `step == last`, then the reader will wait for at least some data to
+    /// arrive, and then create `step` and pass it to the consumer.
+    ///
+    /// - If `step > last`, then the reader will create steps `last..step` with
+    /// any data that has already arrived (possibly none), and then wait for
+    /// some more data to arrive to populate `step`.  It will pass each of these
+    /// to the consumer in turn.
+    ///
+    /// - `step < first` will signal an error through [`InputConsumer::error`].
+    ///
+    /// Each call to `next` must specify a `step` at least as large as the
+    /// previous call.
+    fn next(&self, step: Step);
+}
+
+/// Input consumer for [`DurableInputReader`].
+///
+/// This extension to [`InputConsumer`] divides the input data into individual
```

Comment:
this explanation is pretty abstract, an example for me would be helpful to grasp this

Reply:
OK, I added an example:

```patch
diff --git a/crates/adapters/src/transport/durable/mod.rs b/crates/adapters/src/transport/durable/mod.rs
index 7f897880..9fc9908d 100644
--- a/crates/adapters/src/transport/durable/mod.rs
+++ b/crates/adapters/src/transport/durable/mod.rs
@@ -32,6 +32,33 @@
 //!
 //! [`InputTransport`]: crate::transport::InputTransport
 //! [`InputEndpoint`]: crate::transport::InputEndpoint
+//!
+//! # Example
+//!
+//! ```ignore
+//! let transport = <dyn DurableInputTransport>::get_transport("kafka").unwrap();
+//!
+//! let config_str = format!(
+//!     r#"
+//! data_topic: example_topic
+//! partition: 0
+//! "#
+//! );
+//!
+//! let endpoint = transport
+//!     .new_endpoint(&serde_yaml::from_str(&config_str).unwrap())
+//!     .unwrap();
+//!
+//! let consumer = /*...create a consumer...*/;
+//! let reader = endpoint.open(consumer);
+//! reader.next(5);
+//! // `reader` will asynchronously read step 5 and deliver it to `consumer`
+//!
+//! ...
+//!
+//! reader.next(10);
+//! // `reader` will asynchronously read steps 5..=10 and deliver them to `consumer`
+//! ```
 
 use anyhow::Result as AnyResult;
 use once_cell::sync::Lazy;
```